phoenix-core/src/main/java/org/apache/phoenix/cache/TenantCacheImpl.java

```diff
@@ -77,57 +153,164 @@ public MemoryManager getMemoryManager() {
         return memoryManager;
     }
 
-    private Cache<ImmutableBytesPtr,Closeable> getServerCaches() {
+    private Cache<ImmutableBytesPtr,CacheEntry> getServerCaches() {
         /* Delay creation of this map until it's needed */
         if (serverCaches == null) {
             synchronized(this) {
                 if (serverCaches == null) {
-                    serverCaches = CacheBuilder.newBuilder()
-                        .expireAfterAccess(maxTimeToLiveMs, TimeUnit.MILLISECONDS)
-                        .ticker(getTicker())
-                        .removalListener(new RemovalListener<ImmutableBytesPtr, Closeable>(){
-                            @Override
-                            public void onRemoval(RemovalNotification<ImmutableBytesPtr, Closeable> notification) {
-                                Closeables.closeAllQuietly(Collections.singletonList(notification.getValue()));
-                            }
-                        })
-                        .build();
+                    serverCaches = buildCache(maxTimeToLiveMs, false);
                 }
             }
         }
         return serverCaches;
     }
+
+    private Cache<ImmutableBytesPtr,CacheEntry> getPersistentServerCaches() {
+        /* Delay creation of this map until it's needed */
+        if (persistentServerCaches == null) {
+            synchronized(this) {
+                if (persistentServerCaches == null) {
+                    persistentServerCaches = buildCache(maxPersistenceTimeToLiveMs, true);
+                }
+            }
+        }
+        return persistentServerCaches;
+    }
+
+    private Cache<ImmutableBytesPtr, CacheEntry> buildCache(final int ttl, final boolean isPersistent) {
+        return CacheBuilder.newBuilder()
+            .expireAfterAccess(ttl, TimeUnit.MILLISECONDS)
+            .ticker(getTicker())
+            .removalListener(new RemovalListener<ImmutableBytesPtr, CacheEntry>(){
+                @Override
+                public void onRemoval(RemovalNotification<ImmutableBytesPtr, CacheEntry> notification) {
+                	if (isPersistent || !notification.getValue().getUsePersistentCache()) {
+                        Closeables.closeAllQuietly(Collections.singletonList(notification.getValue()));
```

Comment:
Looks like the indentation is not right here. Can you perform a check through the entire PR?

Reply:
Fixed it here. FWIW some files mix tabs / spaces